### PR TITLE
Add RetinaNet detector

### PR DESCRIPTION
**Pull request by Can-Zhao**
Fixes #.

### Description
Add RetinaNet detector

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/detection/networks/retinanet_detector.py [hunk 1]
```diff
@@ -0,0 +1,896 @@
...
+        self.target_box_key = "boxes"
+        self.target_label_key = "labels"
+        self.pred_score_key = self.target_label_key + "_scores"  # score key for the detected boxes
+
+        # default setting for inference
+        self.use_inferer = False
+        self.score_thresh = 0.5
+        self.topk_candidates = 1000
+        self.nms_thresh = 0.2
+        self.detections_per_img = 300
```
Comment (mibaumgartner) [inline monai/apps/detection/networks/retinanet_detector.py:220]: Since these can be task dependent it would make sense to make them configurable from the outside

File: monai/apps/detection/networks/retinanet_detector.py [hunk 2]
```diff
@@ -0,0 +1,896 @@
...
+                * self.size_divisible[axis]
+            )
+            for axis in range(self.spatial_dims)
+        ]
+
+        # allocate memory for the padded images
+        images = torch.zeros([len(image_sizes), in_channels] + max_spatial_size, dtype=dtype, device=device)
+
+        # Use `SpatialPad` to match sizes, padding in the end will not affect boxes
+        padder = SpatialPad(spatial_size=max_spatial_size, method="end", mode="edge")
```
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/networks/retinanet_detector.py]: Thank you! You are right. I changed it to constant padding

File: monai/apps/detection/networks/retinanet_detector.py [hunk 3]
```diff
@@ -0,0 +1,896 @@
...
+            if self.fg_bg_sampler is None and self.debug:
+                warnings.warn(
+                    "No balanced sampler is used. Negative samples are likely to "
+                    "be much more than positive samples. Please set balanced samplers with self.set_balanced_sampler(*) "
+                    "and self.set_hard_negative_sampler(*), "
+                    "or set classification loss function as Focal loss with self.set_cls_loss(*)"
+                )
+        return
+
+    def preprocess_images(self, input_images: Union[List[Tensor], Tensor]) -> Tuple[Tensor, List[List[int]]]:
```
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/networks/retinanet_detector.py:431]: I added checking, if there no need to pad, will return original tensor

File: monai/apps/detection/networks/retinanet_detector.py (hunk 1 above)
Comment (Can-Zhao) [reply monai/apps/detection/networks/retinanet_detector.py]: Yes, there is such function. I added comments

File: monai/apps/detection/networks/retinanet_detector.py [hunk 4]
```diff
@@ -0,0 +1,896 @@
...
+                # NMS for boxes with label c
+                image_labels_c_idx = (image_labels_t == c).nonzero(as_tuple=False).flatten()
+                keep_c = box_utils.non_max_suppression(
+                    image_boxes_t[image_labels_c_idx, :],
+                    image_scores_t[image_labels_c_idx],
+                    self.nms_thresh,
+                    box_overlap_metric=self.box_overlap_metric,
+                )
+                keep_c = image_labels_c_idx[keep_c[: self.detections_per_img]]  # type: ignore
+                keep.append(keep_c)
```
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/networks/retinanet_detector.py:802]: Thank you! I found it in https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py.
I will add an issue about batch_nms. But leave this version of code as it is for now. I need to make sure the whole detection module can be finished on time.

File: monai/apps/detection/networks/retinanet_detector.py [hunk 5]
```diff
@@ -0,0 +1,923 @@
...
+            detector.network.load_state_dict(torch.load('model.pt'))  # load model
+    """
+
+    def __init__(
+        self,
+        network,
+        anchor_generator: AnchorGenerator,
+        box_overlap_metric: Callable = box_utils.box_iou,
+        debug: bool = False,
+        **kwargs,
```
Comment (Nic-Ma) [inline monai/apps/detection/networks/retinanet_detector.py:176]: missing doc-string for `kwargs`.

Thanks.

File: monai/apps/detection/networks/retinanet_detector.py [hunk 6]
```diff
@@ -0,0 +1,923 @@
...
+from typing import Any, Callable, Dict, List, Sequence, Tuple, Union
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+
+from monai.apps.detection.networks.retinanet_network import RetinaNet, resnet_fpn_feature_extractor
+from monai.apps.detection.utils.anchor_utils import AnchorGenerator
+from monai.apps.detection.utils.box_coder import BoxCoder
+from monai.data import box_utils
```
Comment (Nic-Ma) [inline monai/apps/detection/networks/retinanet_detector.py:51]: I would suggest to import the specific functions instead of the module.

Thanks.

File: monai/apps/detection/networks/retinanet_detector.py [hunk 7]
```diff
@@ -0,0 +1,923 @@
...
+            else:
+                raise ValueError(
+                    "`self.inferer` is not defined."
+                    "Please refer to function self.set_sliding_window_inferer(*) or self.set_custom_inferer(*)."
+                )
+        else:
+            self.use_inferer = False
+        return
+
+    def set_inference_parameters(self, score_thresh=0.05, topk_candidates=1000, nms_thresh=0.5, detections_per_img=300):
```
Comment (Nic-Ma) [inline monai/apps/detection/networks/retinanet_detector.py:317]: Could you please help add typehint for the args?

Thanks.

File: monai/apps/detection/networks/retinanet_detector.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Can-Zhao) [reply monai/apps/detection/networks/retinanet_detector.py]: Thank you so much Michael. I added batch_nms to box_selector in PR https://github.com/Project-MONAI/MONAI/pull/4437

**Review by wyli — APPROVED** [review]
